src/compiler/checker.ts

```diff
@@ -16412,6 +16419,22 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (types.length === 1) {
             return types[0];
         }
+        // We optimize for the common case of unioning a union type with some other type (such as `undefined`).
+        if (types.length === 2 && !origin && (types[0].flags & TypeFlags.Union || types[1].flags & TypeFlags.Union)) {
+            const infix = unionReduction === UnionReduction.None ? "N" : unionReduction === UnionReduction.Subtype ? "S" : "L";
+            const index = types[0].id < types[1].id ? 0 : 1;
+            const id = types[index].id + infix + types[1 - index].id + getAliasId(aliasSymbol, aliasTypeArguments);

```

Comment:
Actually, now that I think about it, does this need to take `origin` into consideration? `getUnionTypeFromSortedList` seems to do a lot more for its cache key and it feels like it's possible to write `number | string` and `string | number` and now have the the way it's printed depend on the order in which the two are evaluated.

Reply:
Oh, nevermind, I just noticed the `origin` check above.

Should we just not pass `origin` to `getUnionTypeWorker` given we know it's undefined? That might help disambiguate things for later reading.